### PR TITLE
Add EE support for PartitionByPlanNode.

### PR DESCRIPTION
**Pull request by BillWhite**
Add EE support for PartionByPlanNode.  The PartitionByPlanNode is a
trivial kind of AggregatePlanNode.  In the EE it's a trivial kind of
AggregateSerialPlanNode.  The only real difference between PartitionByPlanNode
and AggregateSerialPlanNode is that the PartitionByPlanNode knows that
it needs to output one row per input row, and not one per group.  The
algorithm p_execute_tuple knows this, and writes the correct number of
rows.

This supersedes the previous versions of PartitionByPlanNode in Java, and introduces
new EE support.

There is one EE plan execution tests, though more are needed.  This is in PartitionByExecutorTest, though it tests an entire plan.

Finally the facility for testing plans in the EE has been refactored quite a bit to make it easier to use.
There is a function which can be used to verify that output tables have the right values.

### CODE REVIEW

File: src/ee/executors/aggregateexecutor.cpp [hunk 1]
```diff
@@ -505,7 +506,79 @@ class HyperLogLogsToCardAgg : public ApproxCountDistinctAgg {
...
+    // This is the peer count.  It counts the number of
+    // rows whose rank are equal to the last row.  When
+    // we see a row with a new order by expression value,
+    // we reset this to zero.
+    int64_t                            m_peerCount;
+    // This is the last value of the order by expression.  Since the
+    // default order by unit is RANGE, and we don't have any way to
+    // change that syntactically in HSQL, we only get one value for
+    // the order by.  So, we can save this in a single NValue.
+    NValue                             m_lastOrderByValue;
```
Comment (wolffcm) [inline src/ee/executors/aggregateexecutor.cpp]: Historically we have had issues with storing NValues in Aggs due to how we handle inlined strings. See the longish comment in MaxAgg::advance for an explanation.  Consider if this applies here too.  Also, we should have tests for this!

File: src/ee/executors/aggregateexecutor.cpp [hunk 2]
```diff
@@ -505,7 +506,79 @@ class HyperLogLogsToCardAgg : public ApproxCountDistinctAgg {
...
+    // We don't support dense rank now.  But we can easily.  So
+    // add support now.
+    bool                               m_isDenseRank;
```
Comment (wolffcm) [inline src/ee/executors/aggregateexecutor.cpp]: If we are just going to suport dense_rank out of the gate because it's easy, then this comment seems unnecessary?

File: src/ee/executors/partitionbyexecutor.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+#ifndef SRC_EE_EXECUTORS_PARTITIONBYEXECUTOR_H_
+#define SRC_EE_EXECUTORS_PARTITIONBYEXECUTOR_H_
+
+#include "aggregateexecutor.h"
+
+namespace voltdb {
+
+/*
+ *
```
Comment (wolffcm) [inline src/ee/executors/partitionbyexecutor.h]: A Doxygen comment here would be good.

File: src/frontend/org/voltdb/expressions/AbstractExpression.java [hunk 4]
```diff
@@ -709,10 +713,14 @@ private static AbstractExpression fromJSONObject(JSONObject obj,  StmtTableScan
      * The lists are cleared before they are filled in.  This is the inverse of toJSONArrayFromSortList.
      *
      * The JSONObject should be in object state, not array state.  It should have a member
-     * named SORT_COLUMNS, which is an array with the <expression, direction> pairs.
+     * named SORT_COLUMNS, which is an array with the <expression, direction> pairs.  The
+     * directions are optional.  However, if either every expression has a direction or
+     * no expression has a direction and the list sortDirections must be null.
```
Comment (wolffcm) [inline src/frontend/org/voltdb/expressions/AbstractExpression.java]: I'm not following this comment.  If you have `ORDER BY x asc, y desc, z asc` then don't you need the direction list still?  Also, I think you meant to end with "...*then* the list of sortDirections must be null."

File: src/frontend/org/voltdb/expressions/WindowedExpression.java [hunk 5]
```diff
@@ -157,19 +156,24 @@ protected void loadFromJSONObject(JSONObject jobj) throws JSONException {
...
-         * expressions will be serialized in the orderby node which preceeds
-         * the PartitionByPlanNode.
-         */
-        stringer.key(Members.PARTITION_BY_EXPRESSIONS.name()).array();
-        for (int idx = 0; idx < m_partitionByExpressions.size(); idx += 1) {
-            stringer.object();
-            m_partitionByExpressions.get(idx).toJSONString(stringer);
-            stringer.endObject();
+        // Be careful.  This node may have changed to be a
+        // non-windowed expression.  Don't serialize the partition by and sort
```
Comment (wolffcm) [inline src/frontend/org/voltdb/expressions/WindowedExpression.java]: I'm confused by this comment.  Aren't you in the WindowedExpression class here?  How would this node have changed?

File: src/frontend/org/voltdb/planner/AbstractParsedStmt.java [hunk 6]
```diff
@@ -740,6 +762,10 @@ private AbstractExpression parseOperationExpression(VoltXMLElement exprNode) {
         else if (exprType == ExpressionType.OPERATOR_EXISTS) {
             expr = optimizeExistsExpression(expr);
         }
+        if (exprNode.attributes.containsKey("valuetype")) {
+            String valueType = exprNode.attributes.get("valuetype");
```
Comment (wolffcm) [inline src/frontend/org/voltdb/planner/AbstractParsedStmt.java]: What is this for?  Seems unrelated to your changes?

File: src/frontend/org/voltdb/plannodes/AggregatePlanNode.java [hunk 7]
```diff
@@ -287,18 +287,30 @@ public void resolveColumnIndexes()
         resolveColumnIndexesUsingSchema(input_schema);
     }
 
+    /**
+     * Gather up all the expressions whose column indices need to be
```
Comment (wolffcm) [inline src/frontend/org/voltdb/plannodes/AggregatePlanNode.java]: I like this approach and I wonder if it could b extended to other plan nodes as well.

I have a minor issue with the name of the method, since it's not so much that the expressions are resolvable, it's that they absolutely need to be resolved for the plan to be correct.  Maybe `getExpressionsNeedingResolution`?

File: src/frontend/org/voltdb/plannodes/AggregatePlanNode.java [hunk 8]
```diff
@@ -287,18 +287,30 @@ public void resolveColumnIndexes()
...
-        for (SchemaColumn col : m_outputSchema.getColumns()) {
-            output_tves.addAll(ExpressionUtil.getTupleValueExpressions(col.getExpression()));
-        }
+        List<TupleValueExpression> output_tves = getResolvableExpressions();
         for (TupleValueExpression tve : output_tves) {
             int index = tve.resolveColumnIndexesUsingSchema(input_schema);
             if (index == -1) {
                 // check to see if this TVE is the aggregate output
-                // XXX SHOULD MODE THIS STRING TO A STATIC DEF SOMEWHERE
+                // XXX SHOULD MOVE THIS STRING TO A STATIC DEF SOMEWHERE
```
Comment (wolffcm) [inline src/frontend/org/voltdb/plannodes/AggregatePlanNode.java]: You now have this static def, so you could fix this and remove the comment?

File: src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java [hunk 9]
```diff
@@ -26,210 +25,135 @@
...
+                                                            AbstractParsedStmt.TEMP_TABLE_NAME,
+                                                            AbstractParsedStmt.WINDOWED_AGGREGATE_COLUMN_NAME,
+                                                            AbstractParsedStmt.WINDOWED_AGGREGATE_COLUMN_NAME,
+                                                            0);
+        tve.setExpressionType(ExpressionType.VALUE_TUPLE);
+        tve.setValueType(m_windowedExpression.getValueType());
+        tve.setValueSize(m_windowedExpression.getValueSize());
+        // This doesn't really matter, since we will be
+        // generating this.  But it can't be negative.
+        tve.setColumnIndex(0);
```
Comment (wolffcm) [inline src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java]: I would have thought that this would be the windowed expression, instead of a TVE?  What am I missing?

File: src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java [hunk 10]
```diff
@@ -26,210 +25,135 @@
...
+        // This could be the first order by expression.  We currently
+        // just support RANGE units, so there is only one order by expression.
+        // Furthermore, the RANK() operation does not have an argument, unlike,
+        // say, the MEAN(EXP) operation.  So, we pass the only order by
+        // expression in as the aggregate expression.  However, it seems
+        // like propagating this hack out of the EE seems wrong.  So, we
+        // don't add an expression here, and we will add order by expressions
+        // to the plan node, as if we were doing it right.  We will fix it
+        // up, which is to say, we will break it, in the EE.
+        m_aggregateExpressions.add(null);
```
Comment (wolffcm) [inline src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java]: So are you adding null to `m_aggregateExpressions` here as a placeholder for the order by expression you're passing to the RANK agg in the EE?  I don't get why you're appending null to this list here, what am I missing?

File: src/frontend/org/voltdb/types/ExpressionType.java [hunk 11]
```diff
@@ -226,4 +226,24 @@ public boolean isAggregateExpression() {
...
+     * column's value.  This is true when the expression is an aggregate
+     * or a windowed aggregate.  We can't just make the windowed aggregate
+     * operations have class AggregateExpression.class because we will need
+     * the class to create WindowedExpression objects, and these have a
+     * different representation than other aggregate expression objects.
+     * For example, they have a PartitionBy list and an OrderBy list.
+     *
+     * @return true if an expression's value is generated by an early plan node.
+     */
+    public boolean isGeneratedAggregateExpression() {
```
Comment (wolffcm) [inline src/frontend/org/voltdb/types/ExpressionType.java]: I'm having trouble with this method name.  Maybe `isAggregateOrWindowedExpression`?

File: tests/ee/plannodes/PartitionByPlanNodeTest.cpp [hunk 12]
```diff
@@ -89,39 +131,22 @@ class PartitionByPlanNodeTest : public Test {
     }
 };
 
+//
+// There is not much here to test.  The only difference between a
+// PartitionByPlanNode and any other Aggregate node is
```
Comment (wolffcm) [inline tests/ee/plannodes/PartitionByPlanNodeTest.cpp]: Is...?

File: tests/ee/executors/PartitionByExecutorTest.cpp [hunk 13]
```diff
@@ -0,0 +1,672 @@
...
+#include "storage/temptable.h"
+#include "storage/tableutil.h"
+#include "test_utils/plan_testing_baseclass.h"
+#include "test_utils/LoadTableFrom.hpp"
+
+
+namespace {
+const char *plan_strings[] = {
+    //  Plan for this query:
+    //      select ID, A, B, RANK() OVER ( PARTITION BY A ORDER BY B ) from AAA;
```
Comment (wolffcm) [inline tests/ee/executors/PartitionByExecutorTest.cpp]: There's two plans here, what's the statement for the other one?

File: tests/frontend/org/voltdb/planner/PlannerTestCase.java [hunk 14]
```diff
@@ -290,6 +297,36 @@ protected void checkQueriesPlansAreTheSame(String sql1, String sql2) {
         assertEquals(explainStr1, explainStr2);
     }
 
+    protected void verifyOrderByPlanNode(OrderByPlanNode  orderByPlanNode,
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/planner/PlannerTestCase.java]: Would be  good to have a comment on how this method is used, since it may be useful outside of your new test.

File: tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java [hunk 15]
```diff
@@ -0,0 +1,486 @@
...
+        }
+
+        // extra data
+        client.callProcedure("tu.insert", 60, 2);
+        client.callProcedure("tu.insert", 70, 3);
+        client.callProcedure("tu.insert", 80, 2);
+    }
+
+    public void notestNonSupportedCase() throws NoConnectionsException, IOException, ProcCallException {
+        System.out.println("STARTING xin......");
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java]: These are all Xin's tests from his IW branch?

File: tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java [hunk 16]
```diff
@@ -0,0 +1,486 @@
+/* This file is part of VoltDB.
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java]: I see many licenses at the top of this file.

File: tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+                {2, 40, 2}, {2, 50, 3}, {3, 30, 1}, {3, 40, 2}});
+
+        vt = client.callProcedure("@AdHoc", "select b, a, rank() over (partition by b order by a desc) from tm order by b, a;").getResults()[0];
+        validateTableOfLongs(vt, new long[][]{{1, 10, 3},{1, 10, 3}, {1, 20, 2}, {1, 30, 1}, {2, 10, 3},
+                {2, 40, 2}, {2, 50, 1}, {3, 30, 2}, {3, 40, 1}});
+    }
+
+
+    public void notestNoRankScan_NON_UNIQUE() throws NoConnectionsException, IOException, ProcCallException {
+        System.out.println("STARTING xin......");
```
Comment (wolffcm) [inline tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java]: Seems like these tests are all disabled.  Are we able to run any of them yet?

**Comment by wolffcm** [comment]
More tests are coming as part of ENG-10035 I guess?

That's all my review comments for this round.  In summary, I think the inline varchar issue in the Agg subclass in the EE could be serious.  Also, I'd like to understand the issue with passing order by exprs to the rank agg in the EE better.

File: src/ee/executors/aggregateexecutor.cpp (hunk 1 above)
Comment (BillWhite) [inline src/ee/executors/aggregateexecutor.cpp]: These can only be integer or timestamp.  So, I don't think they can be out of line.  When we support ROW units this will have to change.  But I think for now it's ok.

File: src/ee/executors/aggregateexecutor.cpp (hunk 2 above)
Comment (BillWhite) [inline src/ee/executors/aggregateexecutor.cpp]: I didn't want to do all the support right now.  I'd like to get this merged to master and then do the (minimal) changes for dense rank.

File: src/ee/executors/partitionbyexecutor.h (hunk 3 above)
Comment (BillWhite) [inline src/ee/executors/partitionbyexecutor.h]: Done.

File: src/frontend/org/voltdb/expressions/AbstractExpression.java (hunk 4 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/expressions/AbstractExpression.java]: No, we don't need the sort directions.  We need them in the OrderByExecutor for the windowed expression, but
that's earlier in the execution order from us.  By the time the data gets to us it's been sorted.  We need the order by expressions to calculate whether we are in the same order-by peer group.

File: src/frontend/org/voltdb/expressions/WindowedExpression.java (hunk 5 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/expressions/WindowedExpression.java]: %%% 
This is a note to myself, not a comment.  I got to here and then I had to leave.

Comment (BillWhite) [inline src/frontend/org/voltdb/expressions/WindowedExpression.java]: You are right.  This is a comment explaining a way I was doing this before.  I tried making the Windowed Expression look like a TVE.  I decided what I really wanted in the column was a TVE to begin with.  So, the Windowed Expression is never actually put into the plan.  It's cached in the AbstractParsedStmt.  I wanted to cache it in the ParsedSelectStmt, which seems logical.  But the method parseRankExpression is in the AbstractParsedStmt, and it's hard to move it from there.

File: src/frontend/org/voltdb/planner/AbstractParsedStmt.java (hunk 6 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/planner/AbstractParsedStmt.java]: Right.  I was trying to propagate the HSQL from XML to the AbstractParsedStmt.  This is better done in finalizeType, which I now call in the right place.

File: src/frontend/org/voltdb/plannodes/AggregatePlanNode.java (hunk 7 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/plannodes/AggregatePlanNode.java]: Ok.  I'm easy.

File: src/frontend/org/voltdb/plannodes/AggregatePlanNode.java (hunk 8 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/plannodes/AggregatePlanNode.java]: Yes.

File: src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java (hunk 9 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java]: It turns out that putting the windowed expression here causes problems.  This in the output schema, so this is describing the columns of the table we are writing with this partitionby plan node.  Consequently we just want to describe the generated column here.  We don't really care that it's a windowed expression.  We just want to know its type and size for the next plan node in the plan.

File: src/ee/executors/aggregateexecutor.cpp (hunk 1 above)
Comment (wolffcm) [inline src/ee/executors/aggregateexecutor.cpp]: In that case I suggest an assertion to ensure we don't forget.

File: src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java (hunk 10 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/plannodes/PartitionByPlanNode.java]: It turns out that the m_aggregateExpressions list and the m_aggregateTypes list are parallel.  I need to add something here.  It gets fixed up later on in the EE.  This is just to avoid making this look wrong in the planner.  The hack with expressions and orderby is entirely kept in the EE.

File: src/frontend/org/voltdb/types/ExpressionType.java (hunk 11 above)
Comment (BillWhite) [inline src/frontend/org/voltdb/types/ExpressionType.java]: Actually, this bit of code is not necessary.  The WindowedExpression will never be in the display list or the output schema.  It will always be a TVE.

File: tests/ee/executors/PartitionByExecutorTest.cpp (hunk 13 above)
Comment (BillWhite) [inline tests/ee/executors/PartitionByExecutorTest.cpp]: I should put it here.  But this will all be replaced by an automagically generated test case anyway, so I'm inclined to leave it as it is.

File: tests/ee/plannodes/PartitionByPlanNodeTest.cpp (hunk 12 above)
Comment (BillWhite) [inline tests/ee/plannodes/PartitionByPlanNodeTest.cpp]: Fixed.  The PartitionByPlanNode knows how to generate one output row per input row.  Even this is not completely right.  We need to generate N output rows for each order by peer group of size N.  We don't in general know how to do that until we see the entire order by peer group.

File: tests/frontend/org/voltdb/planner/PlannerTestCase.java (hunk 14 above)
Comment (BillWhite) [inline tests/frontend/org/voltdb/planner/PlannerTestCase.java]: Done.

File: tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java (hunk 16 above)
Comment (BillWhite) [inline tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java]: I eliminated this file entirely.  It was mostly an artifact of Xin's implementation, which does not work for us at all.  In particular, there are a lot of rank expressions in where clauses and having clauses, which we don't have a prayer of supporting for quite a long time.

File: tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java (hunk 15 above)
Comment (BillWhite) [inline tests/frontend/org/voltdb/regressionsuites/TestAnalyticRank.java]: Yes.  The ones that make sense were moved to TestWindowedAggregateSuite.

File: src/ee/executors/aggregateexecutor.cpp (hunk 1 above)
Comment (BillWhite) [inline src/ee/executors/aggregateexecutor.cpp]: Actually, I'm changing it to be an int64_t.  Then we have to figure it out eventually.

**Comment by BillWhite** [comment]
Squash merged.
